### ISSUE TITLE
Can't commit in magit

### ISSUE THREAD

**Issue by PythonNut** [Wilfred/remacs#142]
(followup to #92)

When trying to commit staged changes in `magit`, the following error is produced:

> Nothing staged (or unstaged)

In regular Emacs, this works correctly.

Unfortunately, I don't currently have a guess as to why this happens.

**Comment by vodik** [comment]
A hint with this one is the assert looking for staged files fails in remacs which succeeds in emacs. More specifically:

```elisp
(magit-git-exit-code "diff" "--quiet" "--cached")
```

returns 1 under emacs and 0 under remacs.

**Comment by vodik** [comment]
And I think I know what it is. Try this in remacs:

```elisp
(process-file "false")
```

You incorrectly get back 0. Should be 1. Works like that in emacs.

**Comment by vodik** [comment]
And looks like `call-process` is original C code, so don't have the faintest idea how a regression here was introduced.

**Comment by vodik** [comment]
Copying callproc.c from upstream emacs and fixing it up so it would compile resulted in a correction of this problem. I'll investigate a bit further tomorrow, but at the very least I have a fix:

```diff
diff --git a/src/callproc.c b/src/callproc.c
index 113d3f41f6..873065027d 100644
--- a/src/callproc.c
+++ b/src/callproc.c
@@ -182,11 +182,11 @@ call_process_cleanup (Lisp_Object buffer)
     {
       kill (-synch_process_pid, SIGINT);
       message1 ("Waiting for process to die...(type C-g again to kill it instantly)");
-      immediate_quit = true;
-      maybe_quit ();
+
+      /* This will quit on C-g.  */
       wait_for_termination (synch_process_pid, 0, 1);
+
       synch_process_pid = 0;
-      immediate_quit = false;
       message1 ("Waiting for process to die...done");
     }
 }
@@ -293,6 +293,11 @@ call_process (ptrdiff_t nargs, Lisp_Object *args, int filefd,
 
   error_file = Qt;
 
+  /* Without asynchronous processes we cannot have BUFFER == 0.  */
+  if (nargs >= 3
+      && (INTEGERP (CONSP (args[2]) ? XCAR (args[2]) : args[2])))
+    error ("Operating system cannot handle asynchronous subprocesses");
+
   /* Decide the coding-system for giving arguments.  */
   {
     Lisp_Object val, *args2;
@@ -635,9 +640,6 @@ call_process (ptrdiff_t nargs, Lisp_Object *args, int filefd,
       process_coding.src_multibyte = 0;
     }
 
-  immediate_quit = true;
-  maybe_quit ();
-
   if (0 <= fd0)
     {
       enum { CALLPROC_BUFFER_SIZE_MIN = 16 * 1024 };
@@ -659,7 +661,7 @@ call_process (ptrdiff_t nargs, Lisp_Object *args, int filefd,
 	  while (nread < bufsize - 1024)
 	    {
 	      int this_read = emacs_read (fd0, buf + nread,
-					  bufsize - nread);
+					       bufsize - nread);
 
 	      if (this_read < 0)
 		goto give_up;
@@ -678,7 +680,6 @@ call_process (ptrdiff_t nargs, Lisp_Object *args, int filefd,
 	    }
 
 	  /* Now NREAD is the total amount of data in the buffer.  */
-	  immediate_quit = false;
 
 	  if (!nread)
 	    ;
@@ -751,8 +752,6 @@ call_process (ptrdiff_t nargs, Lisp_Object *args, int filefd,
 		 we should have already detected a coding system.  */
 	      display_on_the_fly = true;
 	    }
-	  immediate_quit = true;
-	  maybe_quit ();
 	}
     give_up: ;
 
@@ -764,7 +763,8 @@ call_process (ptrdiff_t nargs, Lisp_Object *args, int filefd,
 	       make_number (total_read));
     }
 
-  immediate_quit = false;
+  /* Wait for it to terminate, unless it already has.  */
+  wait_for_termination (pid, &status, fd0 < 0);
 
   /* Don't kill any children that the subprocess may have left behind
      when exiting.  */
@@ -1228,28 +1228,6 @@ child_setup (int in, int out, int err, char **new_argv, bool set_pgrp,
 #endif  /* not WINDOWSNT */
 }
 
-#ifndef WINDOWSNT
-/* Move the file descriptor FD so that its number is not less than MINFD.
-   If the file descriptor is moved at all, the original is closed on MSDOS,
-   but not elsewhere as the caller will close it anyway.  */
-static int
-relocate_fd (int fd, int minfd)
-{
-  if (fd >= minfd)
-    return fd;
-  else
-    {
-      int new = fcntl (fd, F_DUPFD_CLOEXEC, minfd);
-      if (new == -1)
-	{
-	  emacs_perror ("while setting up child");
-	  _exit (EXIT_CANCELED);
-	}
-      return new;
-    }
-}
-#endif /* not WINDOWSNT */
-
 static bool
 getenv_internal_1 (const char *var, ptrdiff_t varlen, char **value,
 		   ptrdiff_t *valuelen, Lisp_Object env)
```

I wonder if the `wait_for_termination` at the end is the only really important bit missing, but hoping I can get some feedback on this so I can do up a patch.

At least this presents a workaround, magit worked.

**Comment by c-nixon** [comment]
@vodik Thanks for looking into this! It does indeed look like another bug from the concurrency branch, upstream are still working out the wrinkles

**Comment by Wilfred** [comment]
Should be fixed in #158. Do let me know if not.

**Comment by PythonNut** [comment]
@Wilfred yes, it does seem to be fixed. Thanks!
